crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -832,6 +840,228 @@ impl From<ProviderError> for EthFilterError {
+struct CachedMode<
+    Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+> {
+    filter_inner: Arc<EthFilterInner<Eth>>,
+    headers: Vec<(usize, <Eth::Provider as HeaderProvider>::Header)>,
+    current_idx: usize,
+}
+
+/// Mode for processing blocks using range queries for older blocks
+struct RangeBlockMode<
+    Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+> {
+    filter_inner: Arc<EthFilterInner<Eth>>,
+    block_range_iter: BlockRangeInclusiveIter,
+    current_headers: Option<Vec<<Eth::Provider as HeaderProvider>::Header>>,
+    current_header_idx: usize,
+}
+
+impl<
+        Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+    > RangeMode<Eth>
+{
+    /// Creates a new `RangeMode`.
+    fn new(
+        filter_inner: Arc<EthFilterInner<Eth>>,
+        headers: Vec<(usize, <Eth::Provider as HeaderProvider>::Header)>,
+        from_block: u64,
+        to_block: u64,
+        max_headers_range: u64,
+        chain_tip: u64,
+    ) -> Self {
+        let block_count = to_block - from_block + 1;
+        let distance_from_tip = chain_tip.saturating_sub(to_block);
+
+        // use cached mode for recent blocks (close to chain tip) and small ranges
+        if block_count <= CACHED_MODE_BLOCK_THRESHOLD &&
+            distance_from_tip <= CACHED_MODE_BLOCK_THRESHOLD &&
+            !headers.is_empty()
+        {
+            Self::Cached(CachedMode { filter_inner, headers, current_idx: 0 })
+        } else {
+            Self::Range(RangeBlockMode {
+                filter_inner,
+                block_range_iter: BlockRangeInclusiveIter::new(
+                    from_block..=to_block,
+                    max_headers_range,
+                ),
+                current_headers: None,
+                current_header_idx: 0,
+            })
+        }
+    }
+
+    /// Gets the next (receipts, `maybe_block`, header, `block_hash`) tuple.
+    async fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            Arc<Vec<ProviderReceipt<Eth::Provider>>>,
+            Option<Arc<reth_primitives_traits::RecoveredBlock<ProviderBlock<Eth::Provider>>>>,
+            <Eth::Provider as HeaderProvider>::Header,
+            B256,
+        )>,
+        EthFilterError,
+    > {
+        match self {
+            Self::Cached(cached) => cached.next().await,
+            Self::Range(range) => range.next().await,
+        }
+    }
+}
+
+impl<
+        Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+    > CachedMode<Eth>
+{
+    async fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            Arc<Vec<ProviderReceipt<Eth::Provider>>>,
+            Option<Arc<reth_primitives_traits::RecoveredBlock<ProviderBlock<Eth::Provider>>>>,
+            <Eth::Provider as HeaderProvider>::Header,
+            B256,
```

Comment:
we have this a few times, I think worth introducing a helper struct for this?

Reply:
it was an alias, proper struct now